### PR TITLE
New message MSG_EXT_EVENT for reporting timestamped external pin transitions

### PR DESCRIPTION
**Pull request by henryhallam**
Untested at the moment, since I can't get the generator to work.
@denniszollo @fnoble thoughts on the message structure appreciated.  Also, the message ID - does it make sense to make this independently maskable, or should it just be in the same bunch as e.g. the navigation messages?

What should the behavior be if we detect an event but don't yet have a navigation solution and hence can't convert to GPS time?  Perhaps output the time as all zeros, and/or use a flag bit?  Alternately, just don't output messages in that situation.

We should potentially also add an option to generate and output an interpolated RTK position, but I think that would be a separate message.

### CODE REVIEW

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+            desc: GPS week number
+        - tow:
+            type: u32
+            units: ms
+            desc: GPS Time of Week rounded to the nearest ms
+        - ns:
+            type: s32
+            units: ns
+            desc: Nanosecond remainder of rounded tow
+        - pin_edge:
```
Comment (fnoble) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: rename to `flags` as this now contains more than just the pin edge?

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+            type: u8
```
Comment (fnoble) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: Consider making this a `u16` to give more room for future expansion?

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 3]
```diff
@@ -0,0 +1,48 @@
+# Copyright (C) 2015 Swift Navigation Inc.
+# Contact: Fergus Noble <fergus@swiftnav.com>
```
Comment (fnoble) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: I didn't write this! :)

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+#
+# This source is subject to the license found in the file 'LICENSE' which must
+# be be distributed together with this source. All other rights reserved.
+#
+# THIS CODE AND INFORMATION IS PROVIDED "AS IS" WITHOUT WARRANTY OF ANY KIND,
+# EITHER EXPRESSED OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND/OR FITNESS FOR A PARTICULAR PURPOSE.
+
+package: swiftnav.sbp.ext_events
```
Comment (fnoble) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: Perhaps we should make a package called `time`, and move `MSG_GPS_TIME` into there too?

**Comment by fnoble** [comment]
When the change is ready to be merged you should make sure to commit the generated output too.

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+            units: ms
+            desc: GPS Time of Week rounded to the nearest ms
+        - ns:
+            type: s32
+            units: ns
+            desc: Nanosecond remainder of rounded tow
+        - pin_edge:
+            type: u8
+            desc: |
+                Bit 7 (MSB) denotes the new level of the pin, i.e. 1 for
```
Comment (mookerji) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: There's a "syntax" for denoting flags/bitfields that might be useful here.

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+stable: False
+public: True
+include:
+  - types.yaml
+definitions:
+
+ - MSG_EXT_EVENT:
+    id: 0x0300
+    public: True
+    short_desc: External event
```
Comment (mookerji) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: "Reports timestamped, external event"?

File: spec/yaml/swiftnav/sbp/ext_events.yaml (hunk 4 above)
Comment (mookerji) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: That's a good idea, but we may want to discuss it briefly.

File: spec/yaml/swiftnav/sbp/ext_events.yaml (hunk 1 above)
Comment (henryhallam) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: it's the pin and the edge, but sure.. although if there's bitmask syntax that would be better

**Comment by henryhallam** [comment]
@fnoble Now with separate flags and pin number fields.  How do you feel about 0x0300 as message ID?

**Comment by mookerji** [comment]
Now would be a good time to play test this new Makefile that got merged.

**Comment by fnoble** [comment]
@mookerji : I'll let you manage merging this in when you are ready. Do you want to go ahead with changing the package name to `time`?

File: spec/yaml/swiftnav/sbp/ext_events.yaml [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  e.g. camera shutter time
+
+stable: False
+public: True
+include:
+  - types.yaml
+definitions:
+
+ - MSG_EXT_EVENT:
+    id: 0x0300
```
Comment (fnoble) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: Actually - how about making this part of the same grouping as the time message? e.g. `0x0101`? (`GPS_TIME` is `0x0100`)

Comment (henryhallam) [inline spec/yaml/swiftnav/sbp/ext_events.yaml]: I feel like the relationship between ext_events and time isn't really any stronger than e.g. the relationship between navigation and time.  But I don't really care or think it's terribly important.

**Comment by henryhallam** [comment]
Added test, rebased on latest master.  Having trouble building the docs.

**Comment by henryhallam** [comment]
retest this please

**Comment by henryhallam** [comment]
rebased

**Comment by mfine** [comment]
LGTM

**Comment by henryhallam** [comment]
OK I'm going to merge this just for sanity... 'make test' still doesn't work, but I don't think it's related.

**Pull request closed without merging**